sycl/include/sycl/reduction.hpp

```diff
@@ -1288,12 +1328,10 @@ struct NDRangeReduction<
         // This prevents local memory from scaling with elements
         for (int E = 0; E < NElements; ++E) {
 
-          // Copy the element to local memory to prepare it for tree-reduction.
-          LocalReds[LID] = getReducerAccess(Reducer).getElement(E);
-
           typename Reduction::binary_operation BOp;
-          doTreeReduction(WGSize, LID, LocalReds, BOp,
-                          [&]() { NDIt.barrier(); });
+          doTreeReduction<WorkSizeGuarantees::Equal>(
+              WGSize, NDIt, LocalReds, BOp, [&]() { NDIt.barrier(); },
```

Comment:
From https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:what-changed-between:

> The barrier and mem_fence member functions of the nd_item class have been removed. The barrier member function has been replaced by the group_barrier() function, which can be used to synchronize either [work groups](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#work-group) or [sub-groups](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sub-group).

I think all of them might in fact be calling same `workGroupBarrier` (for which we might have a more idiomatic spelling, see the quote above). If so, we can remove this parameter completely.

Can be another PR if you don't want to complicate this one though.

Reply:
Since we're doing refactoring of `doTreeReduction` in this anyway, we may as well do it. Good shout!